File: JBTheatreTools/Tests/JBTheatreToolsTests/PathSafeTests.swift
import XCTest
@testable import JBTheatreTools

/// F8: an API-supplied tag/asset name used as a local filename component must not traverse or hide.
final class PathSafeTests: XCTestCase {
    func testKeepsLegitimateNames() {
        XCTAssertEqual(PathSafe.component("v1.2.3"), "v1.2.3")
        XCTAssertEqual(PathSafe.component("JBTheatreTools-macOS.zip"), "JBTheatreTools-macOS.zip")
        XCTAssertEqual(PathSafe.component("NDITools-Full-macOS-arm64.zip"), "NDITools-Full-macOS-arm64.zip")
    }

    func testNeutralisesTraversalHiddenAndSeparators() {
        for hostile in ["../../Library/LaunchAgents/x", "..\\..\\x", "/etc/passwd", ".ssh", "a/b/c", "a:b"] {
            let out = PathSafe.component(hostile)
            XCTAssertFalse(out.contains("/"), out)
            XCTAssertFalse(out.contains("\\"), out)
            XCTAssertFalse(out.contains(":"), out)
            XCTAssertFalse(out.hasPrefix("."), "must not start with a dot: \(out)")
        }
        XCTAssertEqual(PathSafe.component(""), "_")
        XCTAssertEqual(PathSafe.component("../x"), "_.._x")
    }
}

File: JBTheatreToolsWin/Core.Tests/PathSafeTests.cs
using Xunit;

namespace JBTheatreTools.Tests;

/// <summary>F8: an API-supplied tag/asset name used as a local filename component must not traverse or hide.</summary>
public class PathSafeTests
{
    [Theory]
    [InlineData("v1.2.3", "v1.2.3")]
    [InlineData("JBTheatreTools-macOS.zip", "JBTheatreTools-macOS.zip")]
    [InlineData("PDFTools-Full-macOS-arm64.zip", "PDFTools-Full-macOS-arm64.zip")]
    public void KeepsLegitimateNames(string input, string expected) => Assert.Equal(expected, PathSafe.Component(input));

    [Theory]
    [InlineData("../../Library/x")]
    [InlineData("..\\..\\x")]
    [InlineData("/etc/passwd")]
    [InlineData(".ssh")]
    [InlineData("a/b/c")]
    public void NeutralisesTraversalHiddenAndSeparators(string hostile)
    {
        var o = PathSafe.Component(hostile);
        Assert.DoesNotContain("/", o);
        Assert.DoesNotContain("\\", o);
        Assert.False(o.StartsWith('.'), $"must not start with a dot: {o}");
    }

    [Fact]
    public void EmptyBecomesUnderscore() => Assert.Equal("_", PathSafe.Component(""));
}
